realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -70,7 +72,10 @@
     // thread pool for all async operations (Query & transaction)
     static final RealmThreadPoolExecutor asyncQueryExecutor = RealmThreadPoolExecutor.getInstance();
 
-    protected final List<WeakReference<RealmChangeListener>> changeListeners =
+    protected final Set<RealmChangeListener> changeListeners =
+            new CopyOnWriteArraySet<RealmChangeListener>();
+
+    protected final List<WeakReference<RealmChangeListener>> weakChangeListeners =
```

Comment:
I would add some small comments above both changeListener/weakChangeListener to explain why it is needed.